README.md
# flutter_password_strength

A password strength checker for flutter.

## Features  
  
 - Linear strength indicator.
 - Customise colors, borders, etc.

## Screenshot

![](https://media.giphy.com/media/kbuhBEahzQAtifnfSL/giphy.gif)

## Usage
  ```
    import 'package:flutter_password_strength/flutter_password_strength.dart';

    FlutterPasswordStrength(
      password: _password, 
      strengthCallback: (strength){
        debugPrint(strength.toString());
      }
    )
  ```

### Arguments

| Arguments  | Default  | Type | Description |
| :------------ |---------------:| :---------------| :-----|
| password | required | String | Password  |
| width | null | double | Strength bar width |
| height | 5 | double | Strength bar height |
| strengthColors | null | TweenSequence<Color> | `0.0 ~ 0.25 : red, 0.26 ~ 0.5 : yellow, 0.51 ~ 0.75 : blue, 0.76 ~ 1 : green` |
| backgroundColor | Colors.grey| Color | background for strength bar |
| radius | 0 | double | Strength bar radius  |
| duration | milliseconds: 300 | Duration | Animation duration |
| strengthCallback | null | `void Function(double strength)` | Strength callback, return between 0 to 1 |


CHANGELOG.md
## [0.1.6] - migrate to null safety.

## [0.1.5] - Bugfix: Dispose Issue.

## [0.1.4] - changed license to MIT.

## [0.1.3] - applied flutter format.

## [0.1.2] - added example.

## [0.0.1] - initial release.


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_password_strength_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


test/flutter_password_strength_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_password_strength/flutter_password_strength.dart';

void main() {
  test('calculate password strength', () {
    String password = 'jkdf928n8@0#@ijds!';
    
    FlutterPasswordStrength(password: password, strengthCallback: (strength){
      expect(strength > 0.9, true);
    });
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import './component/sign_in_form.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Password Strength Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Password Strength Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);

  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title!),
      ),
      body: Center(
        child: SignInForm()
      ),
    );
  }
}


example/lib/component/sign_in_form.dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

import 'password_field.dart';
import '../flutter_password_strength/flutter_password_strength.dart';

class SignInForm extends StatefulWidget {
  @override
  _SignInFormState createState() => _SignInFormState();
}

class _SignInFormState extends State<SignInForm> {
  final _formKey = GlobalKey<FormState>();

  String _password = "";

  Widget build(BuildContext context) {
    return Container(
        padding: const EdgeInsets.symmetric(vertical: 8.0, horizontal: 30.0),
        child: Form(
            key: _formKey,
            child: Column(children: <Widget>[
              PasswordField(onChanged: (value) {
                setState(() {
                  _password = value;
                });
              }, onSaved: (value) {
                setState(() {
                  _password = value;
                });
              }),
              Padding(
                  padding: const EdgeInsets.fromLTRB(0, 20, 0, 0),
                  child: FlutterPasswordStrength(
                    password: _password,
                    strengthCallback: (strength) {
                      debugPrint(strength.toString());
                    },
                  )),
              Padding(
                  padding: const EdgeInsets.fromLTRB(0, 20, 0, 0),
                  child: FlutterPasswordStrength(
                      password: _password, width: 100)),
              Padding(
                  padding: const EdgeInsets.fromLTRB(0, 20, 0, 0),
                  child: FlutterPasswordStrength(
                    password: _password,
                    width: 200,
                    height: 30,
                    radius: 15,
                  )),
            ])));
  }
}


example/lib/component/password_field.dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

class PasswordField extends StatefulWidget {
  final Function onSaved;
  final Function onChanged;

  PasswordField({Key? key, required this.onSaved, required this.onChanged})
      : super(key: key);

  @override
  _PasswordFieldState createState() => _PasswordFieldState();
}

class _PasswordFieldState extends State<PasswordField> {

  String? validatePassword(String? value) {
    if (value==null || value.isEmpty) {
      return 'enter password.';
    }

    if (value.length < 4)
      return 'too short.';
    else
      return null;
  }

  Widget build(BuildContext context) {
    return TextFormField(
      keyboardType: TextInputType.text,
      obscureText: true,
      validator: validatePassword,
      autovalidateMode: AutovalidateMode.onUserInteraction,
      onChanged: (value) {
        widget.onChanged(value);
      },
      onSaved: (value) {
        widget.onSaved(value);
      },
      decoration: InputDecoration(
          helperText: 'required',
          labelText: 'Password'),
    );
  }
}


lib/src/estimate_bruteforce_strength.dart
/*
 * Author : Julien Scholz
 * https://github.com/Pikaju/PasswordStrength
 */
import 'dart:math';

double estimateBruteforceStrength(String password) {
  if (password.isEmpty) return 0.0;

  // Check which types of characters are used and create an opinionated bonus.
  double charsetBonus;
  if (RegExp(r'^[a-z]*$').hasMatch(password)) {
    charsetBonus = 1.0;
  } else if (RegExp(r'^[a-z0-9]*$').hasMatch(password)) {
    charsetBonus = 1.2;
  } else if (RegExp(r'^[a-zA-Z]*$').hasMatch(password)) {
    charsetBonus = 1.3;
  } else if (RegExp(r'^[a-z\-_!?]*$').hasMatch(password)) {
    charsetBonus = 1.3;
  } else if (RegExp(r'^[a-zA-Z0-9]*$').hasMatch(password)) {
    charsetBonus = 1.5;
  } else {
    charsetBonus = 1.8;
  }

  final logisticFunction = (double x) {
    return 1.0 / (1.0 + exp(-x));
  };

  final curve = (double x) {
    return logisticFunction((x / 3.0) - 4.0);
  };

  return curve(password.length * charsetBonus);
}


lib/flutter_password_strength.dart
library flutter_password_strength;

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

import 'src/estimate_bruteforce_strength.dart';

class FlutterPasswordStrength extends StatefulWidget {
  final String? password;

  //Strength bar width
  final double? width;

  //Strength bar height
  final double height;

  //Strength bar colors are changed depending on strength
  final Animatable<Color>? strengthColors;

  //Strength bar background color
  final Color backgroundColor;

  //Strength bar radius
  final double radius;

  //Strength bar animation duration
  final Duration? duration;

  //Strength callback
  final void Function(double strength)? strengthCallback;

  const FlutterPasswordStrength(
      { Key? key,
        required this.password,
        this.width,
        this.height = 5,
        this.strengthColors,
        this.backgroundColor = Colors.grey,
        this.radius = 0,
        this.duration,
        this.strengthCallback})
      : super(key: key);

  /*
    default strength bar colors
    This is approximate values
    0.0 ~ 0.25 : red
    0.26 ~ 0.5 : yellow
    0.51 ~ 0.75 : blue
    0.76 ~ 1 : green
  */
  Animatable<Color?> get _strengthColors => (strengthColors != null
      ? strengthColors
      : TweenSequence<Color?>(
          [
            TweenSequenceItem(
              weight: 1.0,
              tween: ColorTween(
                begin: Colors.red,
                end: Colors.yellow,
              ),
            ),
            TweenSequenceItem(
              weight: 1.0,
              tween: ColorTween(
                begin: Colors.yellow,
                end: Colors.blue,
              ),
            ),
            TweenSequenceItem(
              weight: 1.0,
              tween: ColorTween(
                begin: Colors.blue,
                end: Colors.green,
              ),
            ),
          ],
        )) as Animatable<Color?>;

  //default duration is 300 milliseconds
  Duration? get _duration =>
      duration != null ? duration : Duration(milliseconds: 300);

  @override
  _FlutterPasswordStrengthState createState() =>
      _FlutterPasswordStrengthState();
}

class _FlutterPasswordStrengthState extends State<FlutterPasswordStrength>
    with SingleTickerProviderStateMixin {
  //Animation controller for strength bar
  late AnimationController _animationController;

  //Animation for strength bar sharp
  late Animation<double> _strengthBarAnimation;

  //Strength bar colors
  late Animatable<Color?> _strengthBarColors;

  //Strength bar color from the list of strength bar colors
  late Color _strengthBarColor;

  //Strength bar color
  late Color _backgroundColor;

  //Strength bar width
  double? _width;

  //Strength bar height
  late double _height;

  //Strength bar radius, default is 0
  double _radius = 0;

  //Strength callback
  void Function(double strength)? _strengthCallback;

  //_begin is used in _strengthBarAnimation
  double _begin = 0;

  //_end is used in _strengthBarAnimation
  double _end = 0;

  //calculated strength from password
  double _passwordStrength = 0;

  @override
  void initState() {
    super.initState();

    //initialize
    _animationController =
        AnimationController(duration: widget._duration, vsync: this);
    _strengthBarAnimation =
        Tween<double>(begin: _begin, end: _end).animate(_animationController);
    _strengthBarColors = widget._strengthColors;
    _strengthBarColor =
        _strengthBarColors.evaluate(AlwaysStoppedAnimation(_passwordStrength)) ?? Colors.transparent;

    _backgroundColor = widget.backgroundColor;

    _width = widget.width;
    _height = widget.height;
    _radius = widget.radius;
    _strengthCallback = widget.strengthCallback;

    //start animation
    _animationController.forward();
  }

  void animate() {
    //calculate strength
    _passwordStrength = estimateBruteforceStrength(widget.password ?? "");

    _begin = _end;
    _end = _passwordStrength * 100;

    _strengthBarAnimation =
        Tween<double>(begin: _begin, end: _end).animate(_animationController);
    _strengthBarColor =
        _strengthBarColors.evaluate(AlwaysStoppedAnimation(_passwordStrength)) ?? Colors.transparent;

    _animationController.forward(from: 0.0);

    if(_strengthCallback != null){
      _strengthCallback!(_passwordStrength);
    }
  }

  @override
  void dispose() {
    super.dispose();
    _animationController.dispose();
  }

  @override
  void didUpdateWidget(FlutterPasswordStrength oldWidget) {
    super.didUpdateWidget(oldWidget);

    if (oldWidget.password != widget.password) {
      animate();
    }
  }

  @override
  Widget build(BuildContext context) {
    return StrengthBarContainer(
        barColor: _strengthBarColor,
        backgroundColor: _backgroundColor,
        width: _width,
        height: _height,
        radius: _radius,
        animation: _strengthBarAnimation);
  }
}

class StrengthBarContainer extends AnimatedWidget {
  final Color barColor;
  final Color backgroundColor;
  final double? width;
  final double height;
  final double radius;

  const StrengthBarContainer(
      {Key? key,
      required this.barColor,
      required this.backgroundColor,
      this.width,
      required this.height,
      required this.radius,
      required Animation animation})
      : super(key: key, listenable: animation);

  Animation<double> get _percent {
    return listenable as Animation<double>;
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
        builder: (BuildContext context, BoxConstraints constraints) {
      return Container(
          child: CustomPaint(
              size: Size(width ?? constraints.maxWidth, height),
              painter: StrengthBarBackground(
                  backgroundColor: backgroundColor, backgroundRadius: radius),
              foregroundPainter: StrengthBar(
                  barColor: barColor,
                  barRadius: radius,
                  percent: _percent.value)));
    });
  }
}

class StrengthBar extends CustomPainter {
  Color barColor;
  double barRadius;
  double percent;

  StrengthBar({required this.barColor, required this.barRadius, required this.percent});

  @override
  void paint(Canvas canvas, Size size) {
    drawBar(canvas, size);
  }

  void drawBar(Canvas canvas, Size size) {
    Paint paint = Paint()
      ..color = barColor
      ..style = PaintingStyle.fill
      ..strokeCap = StrokeCap.round;

    double left = 0;
    double top = 0;
    double right = size.width / 100 * percent;
    double bottom = size.height;

    //the bar width needs to be bigger than radius width
    if (barRadius != 0 && right > 0 && barRadius * 2 > right) {
      right = barRadius * 2;
    }

    canvas.drawRRect(
        RRect.fromRectAndRadius(
          Rect.fromLTRB(left, top, right, bottom),
          Radius.circular(barRadius),
        ),
        paint);
  }

  @override
  bool shouldRepaint(StrengthBar old) {
    return old.percent != percent;
  }
}

class StrengthBarBackground extends CustomPainter {
  Color backgroundColor;
  double? backgroundRadius;

  StrengthBarBackground({required this.backgroundColor, this.backgroundRadius});

  @override
  void paint(Canvas canvas, Size size) {
    drawBarBackground(canvas, size);
  }

  void drawBarBackground(Canvas canvas, Size size) {
    Paint paint = Paint()
      ..color = backgroundColor
      ..style = PaintingStyle.fill
      ..strokeCap = StrokeCap.round;

    double left = 0;
    double top = 0;
    double right = size.width;
    double bottom = size.height;

    canvas.drawRRect(
        RRect.fromRectAndRadius(
          Rect.fromLTRB(left, top, right, bottom),
          Radius.circular(backgroundRadius ?? 0),
        ),
        paint);
  }

  @override
  bool shouldRepaint(StrengthBarBackground old) {
    return true;
  }
}